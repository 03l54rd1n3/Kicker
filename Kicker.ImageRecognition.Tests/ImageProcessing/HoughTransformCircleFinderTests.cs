using System.Drawing;
using Kicker.ImageRecognition.ImageProcessing;
using Kicker.ImageRecognition.Imaging;
using Kicker.Shared;
using Snapshooter.Xunit;
using Color = Kicker.ImageRecognition.Imaging.Color;

namespace Kicker.ImageRecognition.Tests.ImageProcessing;

public class HoughTransformCircleFinderTests
{
    [Fact]
    public void EnumerateEdges_Snapshot_ReturnsCorrectEdges()
    {
        // Arrage
        var bitmap = (Bitmap) Bitmap.FromFile(@"C:\Users\mmertens\OneDrive - cleverbridge.com\Desktop\edge_detection_test.png");
        using var wrappedBitmap = new BitmapWrapper(bitmap);
        using var directBitmap = new DirectBitmap((short) bitmap.Width, (short) bitmap.Height);
        CopyTo(wrappedBitmap, directBitmap);

        var homography = GetHomography(directBitmap.Width, directBitmap.Height);
        using var imageProcessor = new ImageProcessor().SetHomography(homography);
        imageProcessor.Image = directBitmap;

        var sut = new HoughTransformCircleFinder(imageProcessor, 20);

        // Act
        var returnedPoints = sut.EnumerateEdges(
            0,
            0,
            directBitmap.Width,
            directBitmap.Height,
            200).ToArray();

        // Assert
        Assert.All(returnedPoints, _ => Assert.Equal(Color.White, directBitmap.GetPixel(_.X, _.Y)));
        Snapshot.Match(returnedPoints);

        /*
        var edgeBitmap = new DirectBitmap(wrappedBitmap.Width, wrappedBitmap.Height);
        for (short y = 0; y < edgeBitmap.Height; y++)
        for (short x = 0; x < edgeBitmap.Width; x++)
        {
            var color = returnedPoints.Contains(new PointS(x, y)) ? Color.White : Color.Black;
            edgeBitmap.SetPixel(x, y, color);
        }
        edgeBitmap.Bitmap.Save(@"C:\Users\mmertens\OneDrive - cleverbridge.com\Desktop\edge_detection_test_result.png", ImageFormat.Png);
        */
    }

    [Fact]
    public void FindCircle_Snapshot_ReturnsCorrectEdges()
    {
        // Arrage
        var bitmap = (Bitmap) Bitmap.FromFile(@"C:\Users\mmertens\OneDrive - cleverbridge.com\Desktop\edge_detection_test.png");
        using var wrappedBitmap = new BitmapWrapper(bitmap);
        using var directBitmap = new DirectBitmap((short) bitmap.Width, (short) bitmap.Height);
        CopyTo(wrappedBitmap, directBitmap);

        var homography = GetHomography(directBitmap.Width, directBitmap.Height);
        using var imageProcessor = new ImageProcessor().SetHomography(homography);
        imageProcessor.Image = directBitmap;

        var sut = new HoughTransformCircleFinder(imageProcessor, 15);

        // Act
        var circleCenter = sut.FindCircle(
            0,
            0,
            directBitmap.Width,
            directBitmap.Height,
            200);

        // Assert
        Snapshot.Match(circleCenter);

        directBitmap.SetPixel(circleCenter.X, circleCenter.Y, Color.Green);
        directBitmap.Bitmap.Save(@"C:\Users\mmertens\OneDrive - cleverbridge.com\Desktop\edge_detection_test_center.png", System.Drawing.Imaging.ImageFormat.Png);
    }

    private void CopyTo(
        IImage src,
        IImage dest)
    {
        for (short y = 0; y < src.Height; y++)
        for (short x = 0; x < src.Width; x++)
            dest.SetPixel(x, y, src.GetPixel(x, y));
    }

    private Homography GetHomography(
        short width,
        short height)
        => new Homography()
            .SetOriginFrame(
                new PointS(0, 0),
                new PointS(unchecked((short) (width - 1)), 0),
                new PointS(unchecked((short) (width - 1)), unchecked((short) (height - 1))),
                new PointS(0, unchecked((short) (height - 1))))
            .SetTargetFrame(150, height)
            .CalculateHomographyMatrix()
            .CalculateTranslationMap();
}