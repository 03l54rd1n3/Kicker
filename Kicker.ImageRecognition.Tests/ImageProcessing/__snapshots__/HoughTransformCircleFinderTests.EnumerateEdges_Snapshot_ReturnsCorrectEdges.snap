[
  {
    "X": 53,
    "Y": 9
  },
  {
    "X": 55,
    "Y": 9
  },
  {
    "X": 49,
    "Y": 10
  },
  {
    "X": 61,
    "Y": 10
  },
  {
    "X": 47,
    "Y": 11
  },
  {
    "X": 63,
    "Y": 11
  },
  {
    "X": 46,
    "Y": 12
  },
  {
    "X": 65,
    "Y": 12
  },
  {
    "X": 45,
    "Y": 13
  },
  {
    "X": 67,
    "Y": 13
  },
  {
    "X": 44,
    "Y": 14
  },
  {
    "X": 68,
    "Y": 14
  },
  {
    "X": 110,
    "Y": 14
  },
  {
    "X": 111,
    "Y": 14
  },
  {
    "X": 113,
    "Y": 14
  },
  {
    "X": 44,
    "Y": 15
  },
  {
    "X": 69,
    "Y": 15
  },
  {
    "X": 108,
    "Y": 15
  },
  {
    "X": 111,
    "Y": 15
  },
  {
    "X": 113,
    "Y": 15
  },
  {
    "X": 43,
    "Y": 16
  },
  {
    "X": 70,
    "Y": 16
  },
  {
    "X": 107,
    "Y": 16
  },
  {
    "X": 111,
    "Y": 16
  },
  {
    "X": 113,
    "Y": 16
  },
  {
    "X": 43,
    "Y": 17
  },
  {
    "X": 71,
    "Y": 17
  },
  {
    "X": 107,
    "Y": 17
  },
  {
    "X": 111,
    "Y": 17
  },
  {
    "X": 113,
    "Y": 17
  },
  {
    "X": 42,
    "Y": 18
  },
  {
    "X": 71,
    "Y": 18
  },
  {
    "X": 107,
    "Y": 18
  },
  {
    "X": 111,
    "Y": 18
  },
  {
    "X": 113,
    "Y": 18
  },
  {
    "X": 41,
    "Y": 19
  },
  {
    "X": 71,
    "Y": 19
  },
  {
    "X": 107,
    "Y": 19
  },
  {
    "X": 111,
    "Y": 19
  },
  {
    "X": 113,
    "Y": 19
  },
  {
    "X": 41,
    "Y": 20
  },
  {
    "X": 71,
    "Y": 20
  },
  {
    "X": 107,
    "Y": 20
  },
  {
    "X": 111,
    "Y": 20
  },
  {
    "X": 113,
    "Y": 20
  },
  {
    "X": 41,
    "Y": 21
  },
  {
    "X": 72,
    "Y": 21
  },
  {
    "X": 107,
    "Y": 21
  },
  {
    "X": 111,
    "Y": 21
  },
  {
    "X": 113,
    "Y": 21
  },
  {
    "X": 41,
    "Y": 22
  },
  {
    "X": 72,
    "Y": 22
  },
  {
    "X": 40,
    "Y": 23
  },
  {
    "X": 72,
    "Y": 23
  },
  {
    "X": 40,
    "Y": 24
  },
  {
    "X": 72,
    "Y": 24
  },
  {
    "X": 40,
    "Y": 25
  },
  {
    "X": 72,
    "Y": 25
  },
  {
    "X": 40,
    "Y": 26
  },
  {
    "X": 72,
    "Y": 26
  },
  {
    "X": 40,
    "Y": 27
  },
  {
    "X": 72,
    "Y": 27
  },
  {
    "X": 40,
    "Y": 28
  },
  {
    "X": 71,
    "Y": 28
  },
  {
    "X": 41,
    "Y": 29
  },
  {
    "X": 71,
    "Y": 29
  },
  {
    "X": 41,
    "Y": 30
  },
  {
    "X": 70,
    "Y": 30
  },
  {
    "X": 41,
    "Y": 31
  },
  {
    "X": 70,
    "Y": 31
  },
  {
    "X": 42,
    "Y": 32
  },
  {
    "X": 69,
    "Y": 32
  },
  {
    "X": 43,
    "Y": 33
  },
  {
    "X": 69,
    "Y": 33
  },
  {
    "X": 44,
    "Y": 34
  },
  {
    "X": 67,
    "Y": 34
  },
  {
    "X": 45,
    "Y": 35
  },
  {
    "X": 65,
    "Y": 35
  },
  {
    "X": 46,
    "Y": 36
  },
  {
    "X": 64,
    "Y": 36
  },
  {
    "X": 48,
    "Y": 37
  },
  {
    "X": 61,
    "Y": 37
  },
  {
    "X": 51,
    "Y": 38
  },
  {
    "X": 56,
    "Y": 38
  },
  {
    "X": 47,
    "Y": 36
  },
  {
    "X": 48,
    "Y": 11
  },
  {
    "X": 49,
    "Y": 37
  },
  {
    "X": 50,
    "Y": 10
  },
  {
    "X": 50,
    "Y": 37
  },
  {
    "X": 51,
    "Y": 10
  },
  {
    "X": 52,
    "Y": 10
  },
  {
    "X": 52,
    "Y": 38
  },
  {
    "X": 53,
    "Y": 38
  },
  {
    "X": 54,
    "Y": 9
  },
  {
    "X": 54,
    "Y": 38
  },
  {
    "X": 55,
    "Y": 38
  },
  {
    "X": 56,
    "Y": 10
  },
  {
    "X": 57,
    "Y": 10
  },
  {
    "X": 57,
    "Y": 37
  },
  {
    "X": 58,
    "Y": 10
  },
  {
    "X": 58,
    "Y": 37
  },
  {
    "X": 59,
    "Y": 10
  },
  {
    "X": 59,
    "Y": 37
  },
  {
    "X": 60,
    "Y": 10
  },
  {
    "X": 60,
    "Y": 37
  },
  {
    "X": 62,
    "Y": 11
  },
  {
    "X": 62,
    "Y": 36
  },
  {
    "X": 63,
    "Y": 36
  },
  {
    "X": 64,
    "Y": 12
  },
  {
    "X": 66,
    "Y": 13
  },
  {
    "X": 66,
    "Y": 34
  },
  {
    "X": 68,
    "Y": 33
  },
  {
    "X": 108,
    "Y": 21
  },
  {
    "X": 109,
    "Y": 15
  },
  {
    "X": 109,
    "Y": 21
  },
  {
    "X": 110,
    "Y": 21
  }
]
